dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/TracesFeature.kt

```diff
@@ -0,0 +1,92 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-2020 Datadog, Inc.
+ */
+
+package com.datadog.android.tracing.internal
+
+import android.content.Context
+import android.os.HandlerThread
+import com.datadog.android.DatadogConfig
+import com.datadog.android.DatadogEndpoint
+import com.datadog.android.core.internal.data.upload.DataUploadHandlerThread
+import com.datadog.android.core.internal.domain.NoOpPersistenceStrategy
+import com.datadog.android.core.internal.domain.PersistenceStrategy
+import com.datadog.android.core.internal.net.DataUploader
+import com.datadog.android.core.internal.net.NoOpDataUploader
+import com.datadog.android.core.internal.net.info.NetworkInfoProvider
+import com.datadog.android.core.internal.system.SystemInfoProvider
+import com.datadog.android.core.internal.time.TimeProvider
+import com.datadog.android.tracing.internal.domain.TracingFileStrategy
+import com.datadog.android.tracing.internal.net.TracesOkHttpUploader
+import datadog.opentracing.DDSpan
+import okhttp3.OkHttpClient
+
+internal object TracesFeature {
+    internal var initialized: Boolean = false
+
+    internal var clientToken: String = ""
+    internal var endpointUrl: String = DatadogEndpoint.TRACES_US
+    internal var serviceName: String = DatadogConfig.DEFAULT_SERVICE_NAME
+
+    internal var persistenceStrategy: PersistenceStrategy<DDSpan> = NoOpPersistenceStrategy()
+    internal var uploader: DataUploader = NoOpDataUploader()
+    internal var uploadHandlerThread: HandlerThread = HandlerThread("NoOp")
+
+    @Suppress("LongParameterList")
+    fun initialize(
+        appContext: Context,
+        config: DatadogConfig.FeatureConfig,
+        okHttpClient: OkHttpClient,
+        networkInfoProvider: NetworkInfoProvider,
+        systemInfoProvider: SystemInfoProvider,
+        timeProvider: TimeProvider
+    ) {
+        if (initialized) {
+            return
+        }
+
+        clientToken = config.clientToken
+        endpointUrl = config.endpointUrl
+        serviceName = config.serviceName
+
+        persistenceStrategy = TracingFileStrategy(appContext, timeProvider)
+        setupUploader(endpointUrl, okHttpClient, networkInfoProvider, systemInfoProvider)
+
+        initialized = true
+    }
+
+    fun stop() {
+        uploadHandlerThread.quitSafely()
```

Comment:
Doing that

Reply:
Just thinking, the initialize and stop are internal and will only be called by the Datadog object, which already safeguards the initialize and stop. Beside stop is only called from our tests, and cannot be called from the user's code.

Reply:
Initialize is not actually internal but Stop is indeed so I don't think it will be a concurrency problem there indeed. I am thinking now if we should make all these dependencies in the Feature classes (singleton classes) volatile, meaning that at the bytecode level they should never be returned in an intermediary state (without calling their constructor). This is a pretty rare scenario but it can happen that if I call "Initialize" from one thread and try to use Datadog SDK from another one...those objects to be in an intermediary state even though they were initialized because Java instantiates the Object first and then calls the constructor on a different line if you don't mark it as volatile.